lib/backend/key.go

```diff
@@ -19,81 +19,138 @@ package backend
 import (
 	"bytes"
 	"fmt"
+	"slices"
 	"strings"
 )
 
 // Key is the unique identifier for an [Item].
-type Key []byte
+type Key struct {
+	components []string
+	s          string
+	exactKey   bool
+	noEnd      bool
+}
 
 // Separator is used as a separator between key parts
 const Separator = '/'
+const SeparatorString = string(Separator)
 
 // NewKey joins parts into path separated by Separator,
 // makes sure path always starts with Separator ("/")
-func NewKey(parts ...string) Key {
-	return internalKey("", parts...)
+func NewKey(components ...string) Key {
+	k := internalKey("", components...)
+	k.exactKey = k.s == SeparatorString || (len(k.s) > 0 && k.s[len(k.s)-1] == Separator)
+	return k
 }
 
 // ExactKey is like Key, except a Separator is appended to the result
 // path of Key. This is to ensure range matching of a path will only
 // math child paths and not other paths that have the resulting path
 // as a prefix.
-func ExactKey(parts ...string) Key {
-	return append(NewKey(parts...), Separator)
+func ExactKey(components ...string) Key {
+	k := NewKey(append(components, "")...)
+	k.exactKey = true
+	return k
+}
+
+func KeyFromString(s string) Key {
+	components := strings.Split(s, SeparatorString)
+	if components[0] == "" && len(components) > 1 {
+		components = components[1:]
+	}
+
+	return Key{
+		components: components,
+		s:          s,
+		exactKey:   s == SeparatorString || (len(s) > 0 && s[len(s)-1] == Separator),
+		noEnd:      s == string(noEnd),
+	}
+}
+
+func (k Key) IsZero() bool {
+	return len(k.components) == 0 && k.s == ""
+}
+
+func internalKey(internalPrefix string, components ...string) Key {
```

Comment:
This function is pretty dense.  I think some detailed comments would be helpful.